usr/sbin/p11sak/p11sak.c

```diff
@@ -2292,11 +2287,47 @@ static CK_RV generate_asymmetric_key(CK_SESSION_HANDLE session, CK_SLOT_ID slot,
         goto done;
     }
 
-    rc = set_battr(attr_string, pub_attr, &pub_acount, prv_attr, &prv_acount);
-    if (rc != CKR_OK) {
-        fprintf(stderr, "Error setting binary attributes (error code 0x%lX: %s)\n", rc,
-                p11_get_ckr(rc));
-        goto done;
+    /**
+     * if the separator sign is not in the string the custom attributes apply for 
+     * public and private key
+     * else the attributes in front of the separator sign are set for the public 
+     * key and the ones behind are set for the private key
+     */
+    char *ret = NULL;
+    if (attr_string) {
+        ret = strchr(attr_string, separator);
+    }
+
+    if (attr_string && ret != NULL) {
+        // setting the attributes after the separator for the private key
+        rc = set_battr(ret + 1, prv_attr, &prv_acount, 1);
+        if (rc != CKR_OK) {
+            fprintf(stderr, "Error setting binary attributes for private key (error code 0x%lX: %s)\n", rc,
+                    p11_get_ckr(rc));
+            goto done;
+        }
+        // setting the attributes in front of the separator for the public key
+        *ret = '\0';
+        rc = set_battr(attr_string, pub_attr, &pub_acount, 0);
+        if (rc != CKR_OK) {
+            fprintf(stderr, "Error setting binary attributes for public key (error code 0x%lX: %s)\n", rc,
+                    p11_get_ckr(rc));
+            goto done;
+        }
+        *ret = separator;
+    } else {
```

Comment:
shouldn't this be `} else if if (attr_string) {` ? You want to do this only if attrs_string is not NULL. Or is set_battr() smart enough to do nothing in case of attr_string is NULL?

Reply:
In case attr_string == NULL th CKA_TOKEN attribute has to be set by set_battr().

Reply:
OK, so maybe for clearer code and to show that you intentionally pass a NULL attr_string to set_battr():
`rc = set_battr(NULL, .....`
(2 places, one for bub_attr, one for prov_attr).

Reply:
But then I don't cover the case in which attr_string is NOT NULL and ret is NULL, in which case the attributes defined in attr_string should apply for both the private and the public key. So I can't pass NULL...

Reply:
Ah yes, right. So leave it as is.